File: client/src/game/components/fixture.lua
local constants = require 'src.game.components.constants'

local Fixture = {
  _meta = constants.Fixture
}

function Fixture.new(id, entity, fixture, category, mask, group)
  local instance = {
    _meta = Fixture._meta,
    id = id,
    entity = entity,
    fixture = fixture
  }

  -- set the entity to the fixture as userData
  local data = fixture:getUserData() or {}
  data.entity = entity
  fixture:setUserData(data)
  fixture:getBody():setUserData(data)

  function instance:destroy()
    self.fixture:getBody():destroy()
  end

  return instance
end

return Fixture

File: client2/src/game/components/fixture.ts
import { Component } from '../../ecs/Component'
import { Entity } from '../../ecs/Entity'
import { Flag } from './flags'

export class EntityFixture extends Component {
  static _id = 'EntityFixture'
  _id = EntityFixture._id

  static _flag = Flag.EntityFixture
  _flag = Flag.EntityFixture

  entity: Entity
  fixture: Fixture

  constructor(entity: Entity, fixture: Fixture) {
    super()

    this.entity = entity
    this.fixture = fixture

    const data = fixture.getUserData() || {}
    data.entity = this.entity
    fixture.setUserData(data)
    fixture.getBody().setUserData(data)
  }

  destroy = () => {
    this.fixture.getBody().destroy()
  }
}
